datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -118,21 +137,86 @@ type CommonExprs = IndexMap<Identifier, (Expr, String)>;
 /// ProjectionExec(exprs=[extract (day from new_col), extract (year from new_col)]) <-- reuse here
 ///   ProjectionExec(exprs=[to_date(c1) as new_col]) <-- compute to_date once
 /// ```
-pub struct CommonSubexprEliminate {}
+pub struct CommonSubexprEliminate {
+    random_state: RandomState,
+}
 
 impl CommonSubexprEliminate {
+    pub fn new() -> Self {
+        Self {
+            random_state: RandomState::new(),
+        }
+    }
+
+    /// Returns the identifier list for each element in `exprs` and a flag to indicate if
+    /// rewrite phase of CSE make sense.
+    ///
+    /// Returns and array with 1 element for each input expr in `exprs`
+    ///
+    /// Each element is itself the result of [`CommonSubexprEliminate::expr_to_identifier`] for that expr
+    /// (e.g. the identifiers for each node in the tree)
+    fn to_arrays<'n>(
+        &self,
+        exprs: &'n [Expr],
+        expr_stats: &mut ExprStats<'n>,
+        expr_mask: ExprMask,
+    ) -> Result<(bool, Vec<IdArray<'n>>)> {
+        let mut found_common = false;
+        exprs
+            .iter()
+            .map(|e| {
+                let mut id_array = vec![];
+                self.expr_to_identifier(e, expr_stats, &mut id_array, expr_mask)
+                    .map(|fc| {
+                        found_common |= fc;
+
+                        id_array
+                    })
+            })
+            .collect::<Result<Vec<_>>>()
+            .map(|id_arrays| (found_common, id_arrays))
+    }
+
+    /// Go through an expression tree and generate identifier for every node in this tree.
```

Comment:
```suggestion
    /// Add an identifier to `id_array` for every subexpression in this tree.
```

Reply:
Fixed in https://github.com/apache/datafusion/commit/ee61224f1a948c38ac99ae53d9d426bb69da9548.